kythe/go/platform/analysis/local/local.go

```
@@ -140,3 +145,99 @@ type kzipFetcher struct{ r *kzip.Reader }
+		return nil, fmt.Errorf("parameter 'cmd' must not be empty")
+	}
+	return &localAnalyzer{cmd: cmd, fetcher: fetcher}, nil
+}
+
+// Analyze implements the analysis.CompilationAnalyzer interface.
+func (a *localAnalyzer) Analyze(ctx context.Context, req *apb.AnalysisRequest, f analysis.OutputFunc) error {
+	tmpkzip, err := saveSingleUnitKzip(ctx, req.Compilation, a.fetcher)
+	if err != nil {
+		return fmt.Errorf("creating tmp kzip: %v", err)
+	}
+	defer os.Remove(tmpkzip)
+
+	// Run indexer binary
+	var errBuf strings.Builder
+	var outBuf bytes.Buffer
+	cmd := exec.Command(a.cmd[0], append(a.cmd[1:], tmpkzip)...)
+	cmd.Stdout = &outBuf
+	cmd.Stderr = &errBuf
+	if err = cmd.Run(); err != nil {
+		return fmt.Errorf("indexing kzip %s: %v. stderr=%s", tmpkzip, err, errBuf.String())
+	}
+
+	// Deserialize output and pass it to `f`.
+	rd := delimited.NewReader(&outBuf)
+	for {
+		n, err := rd.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return fmt.Errorf("parsing entry stream: %v", err)
+		}
+
+		if err := f(ctx, &apb.AnalysisOutput{Value: n}); err != nil {
+			return fmt.Errorf("unable to write analysis output: %v", err)
+		}
+	}
+	if err := f(ctx, &apb.AnalysisOutput{FinalResult: &apb.AnalysisResult{Status: apb.AnalysisResult_COMPLETE}}); err != nil {
+		return fmt.Errorf("unable to write analysis output: %v", err)
+	}
+
+	return nil
+}
+
+// saveSingleUnitKzip creates a kzip file containing the given compilation unit
+// and any files it depends on. File contents are retrieved from the given
+// fetcher object.
+func saveSingleUnitKzip(ctx context.Context, cu *apb.CompilationUnit, fetcher analysis.Fetcher) (string, error) {
+	tmpfile, err := vfs.CreateTempFile(ctx, "", fmt.Sprintf("*_%s.kzip", cu.VName.Language))
+	if err != nil {
+		return "", err
+	}
+
+	w, err := kzip.NewWriter(tmpfile)
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		if err := w.Close(); err != nil {
+			log.Printf("unable to close kzip writer: %v", err)
```

This error should be returned.

---

Good point. What's the best way to do this since there are multiple early returns?

---

One way is to name the error return value:

`func saveSingleUnitKzip(ctx context.Context, cu *apb.CompilationUnit, fetcher analysis.Fetcher) (path string, err error) {`

And set it in the `defer`:

```
	defer func() {
		if closeErr := w.Close(); closeErr != nil {
                        if err != nil {
                            log.Printf("WARNING: unable to close kzip writer: %v", closeErr)
                        } else {
                            err = closeErr
                        }
		}
	}()
```

Or you can `w.Close()` on every early exit, ignoring errors, and only return the `Close` errors on the final `return`.

---

:thumbsup: